### PR TITLE
add new config options for custom javascripts and stylesheets

### LINKED ISSUE #2425: [Feature request]Adding stylesheets / javascripts to the default ones
Adding a custom stylesheet on all pages is a bit cumbersome : you have to copy paste an excerpt of the output of `app/console config:dump-reference SonataAdminBundle` into your `config.yml`. It would be nice to have an `extra_stylesheets` and an `extra_javascripts` option that, when used, would add to the default stylesheets / javascripts instead of replacing them. What do you think ?

### PR DESCRIPTION
**Pull request by OskarStark**
Fixes #2425

### CODE REVIEW

**Comment by DracoBlue** [comment]
Looks sufficient to me. :+1:

**Comment by joaoalves89** [comment]
Looks good :+1:

**Comment by OskarStark** [comment]
renamed ```custom_*``` to ```extra_*```

**Comment by greg0ire** [comment]
> renamed custom_* to extra_*

Thanks!

**Comment by OskarStark** [comment]
@rande @Soullivaneuh please have a look and merge this PR if it's fine!

Thank you!

File: DependencyInjection/SonataAdminExtension.php [hunk 1]
```diff
@@ -85,8 +85,23 @@ public function load(array $configs, ContainerBuilder $container)
         $processor = new Processor();
         $config = $processor->processConfiguration($configuration, $configs);
 
-        $config['options']['javascripts'] = $config['assets']['javascripts'];
-        $config['options']['stylesheets'] = $config['assets']['stylesheets'];
+        $javascripts = array_unique(
+            array_merge(
+                $config['assets']['javascripts'],
+                $config['assets']['extra_javascripts']
```
Comment (greg0ire) [inline DependencyInjection/SonataAdminExtension.php]: What if the `extra_javascripts` configuration is not defined ? Does this somehow default to `[]` ?

Comment (OskarStark) [inline DependencyInjection/SonataAdminExtension.php]: i tested it and and arrayNode() which is not required should be handeled by the configuration itself IMO

but i'll check that again

Comment (OskarStark) [inline DependencyInjection/SonataAdminExtension.php]: yes it works!

Comment (greg0ire) [inline DependencyInjection/SonataAdminExtension.php]: Ok great! I had a doubt…

Comment (OskarStark) [inline DependencyInjection/SonataAdminExtension.php]: :+1:  no problem.

thank you for the review

**Comment by OskarStark** [comment]
added the 2 new parameters to the docs

**Comment by Soullivaneuh** [comment]
I see the goal of this PR but as `stylesheets` is already configurable, I think this is too "complex".

Maybe `stylesheets` should be deleted (or deprecated) but this will annouce another problem: We won't be able to delete unwanted stylesheets. Same for javascripts of course.

@rande what do you think?

**Comment by greg0ire** [comment]
Then maybe there should simply be a boolean indicating whether things are supposed to get merged or replaced ?

**Comment by Soullivaneuh** [comment]
> Then maybe there should simply be a boolean indicating whether things are supposed to get merged or supposed to replace existing things ?

Not sure of what you mean, can you please elaborate?

**Comment by greg0ire** [comment]
I mean something like this : 

```
javascripts:
    - my.js
    - myOther.js

javascripts_mode: append|replace
```

Something like this, with a better name as `javascripts_mode`

**Comment by Soullivaneuh** [comment]
Looks better, but we don't have solution if we just want to removed one of the vendor for example.

**Comment by greg0ire** [comment]
Apart from writing the whole list minus one js you mean ?

**Comment by Soullivaneuh** [comment]
@greg0ire Yes, exactly. :-)

**Comment by greg0ire** [comment]
Ok, other proposal :

```
javascripts:
    clear: false # boolean, default false
    append:
        - my.js
        - myOther.js
    remove:
        - someExisting.js
```

**Comment by OskarStark** [comment]
@Soullivaneuh we actually have the problem that we cannot remove OR add new stylesheets.

this PR fixes to the ```add``` problem and stay BC

the ```remove``` problem is another task, but IMO not really needed. IF these are default values, they ARE needed. SO we have to rethink the solution of adding custom stylsheets.

in my case we can extend the sonata formatter configuration and add ```extra_stylesheets``` to the config, which will then automatically be merged.

@greg0ire the last one is too complex IMO.

otherwise we could load the default files by default without the configuration and use the stylesheet node....

**Comment by OskarStark** [comment]
i think remove a default JS/CSS isn't a good idea...

do have any use cases @Soullivaneuh

**Comment by greg0ire** [comment]
> @greg0ire the last one is too complex IMO.

Yeah, and it also is a BC-break. Not sure why one would want to remove existing javascripts… I agree it is a separate problem, which should have its separate, BC-compatible solution (a `remove_javascripts` configuration key maybe) if it turns out to be really something people ask for.

**Comment by Soullivaneuh** [comment]
> this PR fixe the add problem and stays BC

I agree. But I think it's better to solve all problem properly with one PR instead of adding extra keys.

> Not sure why one would want to remove existing javascripts

Remove `totolib.js` to replace it by another version of it from bower for example.

**Comment by OskarStark** [comment]
> Remove `totolib.js` to replace it by another version of it from bower for example.

You can use as it is now:
1. add all default scripts under ```javascripts```
2. remove the ```toollib.js``` from ```javascripts```
3. add new ```toollib.js``` to ```extra_javascripts```

but after that, you are resonsible to stay up to date and in sync with sonata, and that is ok IMO

if this is really needed by the community we should solve this, but i'm not sure it is

**Comment by Soullivaneuh** [comment]
> add all default scripts under `javascripts`

Yeah, that's exactly the painful way I would like to avoid. :-)

I like the last @greg0ire proposition BTW.

**Comment by OskarStark** [comment]
ok but to stay BC the ```clear``` node is not neccessary.

my proposition:
```yml
javascripts: # the defaults
    - foo.js 
    - bar.js
extra_javascripts:
        - my.js
        - barNew.js
remove_javascripts:
        - bar.js
```

**Comment by Soullivaneuh** [comment]
I still prefer @greg0ire one.

Could simple scalar and sub arrays be both managed with deprecation triggering?

**Comment by OskarStark** [comment]
i'm not sure, lets find out :)

maybe we don't need to deprecate something in this case and let the extension handle it

**Comment by OskarStark** [comment]
but if we have ```append``` and ```remove``` we don't need the ```clear``` node, right?

removing ALL by setting ```clear: true``` is not really useful

**Comment by mvhirsch** [comment]
Another proposal could be the following:
Always add default values and control them with the configuration like this.

```
javascripts:
    - foo.js           # append
    - bar.js: remove   # remove this from defaults
```

This way, you could keep BC, because the Extension would remove the duplicate javascripts (if someone already copied the defaults over to his/hers project and modified them).
If you start a new project, there is just one thing that would change: no need to copy all defaults over and append you javascript. Just add one line and you're ready to go.

**Comment by Soullivaneuh** [comment]
@mvhirsch I don't think your version is YAML compliant.

Futhermore, I don't like the idea of using filename as key, too risky IMHO.

**Comment by mvhirsch** [comment]
@Soullivaneuh it's called [Compact Nested Mapping](http://www.yaml.org/spec/1.2/spec.html#id2759963).
But yes, it would be the key instead of value.

**Comment by OskarStark** [comment]
Well i think we slept a few nights :smile: 

any new ideas?

The first question IMO is: Do we want a BC break and create a "right" solution?

**Comment by Soullivaneuh** [comment]
> Do we want a BC break and create a "right" solution?

I will correct: We want to **deprecate before BC break** and create a "right" solution. :-)

**Comment by OskarStark** [comment]
ok, so i would like to have such solution:

```yml
javascripts:
    append:
        - /bundles/app/my.js
        - /bundles/app/myOther.js
    remove:
        - bundles/sonata/foo/someExisting.js
```

but i dont't like the clear node:
```yml
    clear: false # boolean, default false
```

**Comment by greg0ire** [comment]
Agreed, `clear` is not useful. The rest looks nice and clean.

**Comment by Soullivaneuh** [comment]
I think `clear` is useful. Otherwise, if we want to replace all js/css file, we have to put it all to `remove` and this will not work after update if another library is added.

**Comment by OskarStark** [comment]
But then you have the same problem if sonata admin now provides 10 JS files and you would like to replace all of them, you would use ```clear``` option.

but after that, when we add some features and provide 11 files, you gonna miss it automatically, because YOU define which files to use... like the current solution

**Comment by Soullivaneuh** [comment]
> Because YOU define which files to use...

As you say, YOU define. :-)

Well, maybe we could find a way to automatically warn users about new files. But IMO it's still to developer responsibility. And BTW, it's actually the case with the current config.

**Comment by greg0ire** [comment]
> As you say, YOU define. :-)

Bu who really needs to do that ? I mean redefine all javascripts and stylesheet. If there is a need and we do not add the `clear` key, we can still add it later, there will not be any BC-break. Plus, we will find the rationale behind such a decision in the process (and document it).

**Comment by jvasseur** [comment]
Maybe we can just keep the old style for people wanting to replace all assets instead of having a ```clear``` key.

That would be a lot cleaner IMO, remove key is useless if you clear first and append doesn't really have a lot of sense either.

This would need a bit of work to switch between both syntax but we would need to do it anyway to keep BC.

**Comment by Soullivaneuh** [comment]
> remove key is useless if you clear first and append doesn't really have a lot of sense either.

Different cases, different needs. This is why I suggest to implement both of them. :+1:

**Comment by OskarStark** [comment]
@rande if you have a few minutes, could you please join this conversation?

**Comment by OskarStark** [comment]
what about something like this:

```php
                        ->scalarNode('javascripts')
                            ->beforeNormalization()
                                ->always()
                                ->then(function ($v) {
                                    @trigger_error('The javascripts configuration key is deprecated since 2.4 and will be removed in 3.0.', E_USER_DEPRECATED);
                                    return $v;
                                })
                            ->end()
                        ->end()
```

**Comment by Soullivaneuh** [comment]
@OskarStark as I said, I would prefer keep the actual config key on just change the content.

**Comment by jvasseur** [comment]
I tried to do an alternative implementation (https://github.com/jvasseur/SonataAdminBundle/tree/assets-config) that would allow keeping the old config and supporting append and remove keys with normalization.

The problem is that it doesn't work when merging multiple config files.

**Comment by OskarStark** [comment]
i like your idea @jvasseur, can you provide it as a PR, so we can discuss it directly?

**Comment by jvasseur** [comment]
@OskarStark Done in #3336

**Comment by OskarStark** [comment]
Thank you

**Comment by OskarStark** [comment]
rebased

**Comment by OskarStark** [comment]
@greg0ire what do you think?

**Comment by greg0ire** [comment]
I think we should merge this instead of letting it rot.

**Comment by ju1ius** [comment]
As some of you may already know, the right solution is called [code splitting](http://jonathancreamer.com/advanced-webpack-part-2-code-splitting/). It is supported by all the popular bundlers (webpack, browserify, rollup, ...).
Your app has required dependencies, say jquery, along with optional plugin lambda (with it's associated stylesheet).
You don't let users replace your required dependencies, because that might break your stuff right ?
For optional deps, you pass a config flag to your main script.
```yaml
# config.yml
assets:
  javascripts:
    - '/path/to/userland/custom.js'
plugins:
  lambda:
    enabled: true
    foo: bar
```
```twig
{# standard_layout.html.twig #}
<script>
window.Sonata.config = {
  pluginLambda: {{ sonata_admin.adminPool.getOption('lambda_config')|json_encode }}
};
</script>
<script src="{{ asset('/sonataadmin/main.js') }}"></script>
{% block javascripts %}
  {% for javascript in sonata_admin.adminPool.getOption('javascripts', []) %}
       <script src="{{ asset(javascript) }}"></script>
   {% endfor %}
{% endblock %}
```
And now this is your main js entry point:
```js
// main.js
// hard dependencies, included in the main bundle.
import $ from 'jquery';
import config from 'sonata/config';

$(() => {
  // ... 
});

if (window.Sonata.config.pluginLambda.enabled) {
  // Optional deps, bundled separately & loaded at runtime.
  Promise.all([
    System.import('lambda'),
    System.import('lambda.css'),
  ]).then(([lambda]) => {
    lambda.default();
  });
}
```
Sorry for the long comment.

**Comment by Soullivaneuh** [comment]
According to the new [Sonata version management](https://github.com/sonata-project/SonataAdminBundle/issues/3053) and next major [release plan](https://github.com/sonata-project/SonataAdminBundle/issues/3731), this project has been refactored regarding branching and versioning.

If you see this message, your PR concerns a patch or a minor release and is not targeting the right branch.

So I'm closing this one, but don't see it as a refusal. If you think your work is still relevant and want to continue, feel free to reopen it on the right branch (e.g. the default one).

Regards.

**Comment by akrz** [comment]
Is it possible to merge this?

**Pull request closed without merging**